### ISSUE TITLE
Need suggestions to reduce the "bleeding" of predicted masks

### ISSUE THREAD

**Issue by DanielFennhagen** [facebookresearch/detr#255]
Hi again!

I have had some success using detr, however I run into a issue with the panoptic segmentation which I don't encounter using other models like Mask RCNN.

First lets get some things out of the way:
* The model is fine tuned from your r50 checkpoint.
* It's trained on mostly generated data (i.e. not real images) (however, I don't think it matters and I'll get into that later)
* It's trained with panoptic image data using `dataset_file = "coco_panoptic"` and `coco_panoptic_path = "pano_dir"`
* The model is unchanged except for a modification to `num_classes`

Anyways, let's have a look at this image (which is from the same source as some of the generated training data, however not in the training):
![carimage](https://user-images.githubusercontent.com/70580355/95972634-fb1b4280-0e12-11eb-8bc4-d5f5d5863d4e.png)

Firstly, the prediction is very good, and I would not complain if all my images were this good, but lets ignore that for now.
What I'd like to focus on is the dark blue part at the far right.

We'll get back to this but lets first have a look at what the object detection head has predicted for this image:
![carimage_bbox](https://user-images.githubusercontent.com/70580355/95973887-80ebbd80-0e14-11eb-9d4b-f6bc369e0cc7.png)

There's a lot of bounding boxes here but what I'd like to focus on is the **rear_door**, its clearly not containing the very right parts that we see in the previous image.

Alright, let's now get into a real example where the effect is more potent:
![045](https://user-images.githubusercontent.com/70580355/95981379-d6c56300-0e1e-11eb-81f3-4770055e1026.jpg)

- and masks predicted from this image:
![54 segm](https://user-images.githubusercontent.com/70580355/96463890-7e7bd000-1227-11eb-8884-88e04abc64ae.png)

<details closed>
<summary>Boxes just to make sure nothing is wrong</summary>
<br>
<img src="https://user-images.githubusercontent.com/70580355/96084185-ed22fb80-0ebe-11eb-9ccb-2badfaa42c20.png">

Some labels are overlapping but from experience I can say it's front_door and not_door overlapping.
</details>

The front fender I have no problem with, and I think it's obvious that the rear_door has some issues.
I would like to add this effect is not found in all images, I'd say that half of them are good predictions and half contains this "bleeding" effect.

The question I like to solve is: Why is this occuring in detr and not in other models?
I have trained a Mask RCNN model using the same dataset, and this effect never occurs in that model. Since Mask RCNN persorms well on this datased I'd like to think the problem is not from the dataset. I understand that Mask RCNN's output is limitied by the RoI which makes the masks not "bleed" outside of the detected bounding box, however I thought the transformer would work in a similar way?
<details closed>
<summary>Mask RCNN in case above statement is not believable</summary>
<br>
<img src="https://user-images.githubusercontent.com/70580355/96464348-fc3fdb80-1227-11eb-9d4b-a8b6954da864.jpg">
</details>

When detr predicts without the bleeding, mask are consistently better than our other models. However the bleeding in about half of the images ruin the results.

Have you guys encountered this problem and what was your solution to this?

Thanks!

**Comment by DanielFennhagen** [comment]
Hi @alcinos, thanks for the quick reply!

>You mention that you are using DETR in a panoptic segmentation way, is this correct? 

Yes, I train detr using the coco_panoptic flag and feed the network with panoptic images generated from the object_detection json file. (The network is obviously also fed with normal images along with the object_detection json)
<details closed>
<summary>A example panoptic train image</summary>
<br>
All pixels are colored, and again, this is generated using the segment data from our object_detection json file.
There is a check at generation stage which makes asserts no segments are overlapping.
<img src="https://user-images.githubusercontent.com/70580355/96549128-589a0e00-12af-11eb-9ee2-60934652a8c5.png">
</details>

> If so, I presume you are also using our panoptic post-processor?

At inference we use the post-proccessor.

> Could you clarify if you indeed need the panoptic property ...  or not?

We do not need the panoptic property, however during some testing we found that the masks had clearer boundaries when we trained using the panoptic setting, is this wrong?

> For your red car example, I don't understand why the mask image contains only 3 boxes, while you plot 5 in the box-only image? What happened to the overlapping front_door and not_door boxes?

I believe the front_door mask had too low confidence and was culled from the result.
I ran inference on your [colab notebook](https://colab.research.google.com/github/facebookresearch/detr/blob/colab/notebooks/DETR_panoptic.ipynb) and recieved this result:
![045](https://user-images.githubusercontent.com/70580355/96550533-49b45b00-12b1-11eb-9ddb-bb81bf2e3317.png)

and for the not_door, I remove that mask in my overlay function, sorry for not specifying that.
I have not looked into the post proccessor so I don't know how customizable it is, but just in case I displayed the results before the post proccessor:
<details closed>
<summary>70% and 30% confidence keep</summary>
<br>
<img src="https://user-images.githubusercontent.com/70580355/96553219-212e6000-12b5-11eb-9f3f-d76e53d33b9d.png">
<img src="https://user-images.githubusercontent.com/70580355/96553214-1ecc0600-12b5-11eb-9791-13a8503ce62a.png">
</details>


> Also, it seems that the green car image is synthetic while the red one is real, is it the case that the model suffers from data distribution shift? In that case, I'd advise looking into more aggressive data-augmentation strategies.

This is definitely something to keep in mind, however I believe that the distribution is currently fair considering it other models converge with decent results.

> Finally, last but not least, a bug was recently spotted in the mask head (see #247), it could reduce performance a bit in your case as well.

I'll modify my fork to include the code from #247, thanks!


Thanks again and looking forward to your respone :)

**Comment by alcinos** [comment]
The panoptic post-processor is optimized to maximize the PQ metric, but I don't think that's very align with what you want. In your example, since your front door box is dropped, the remaining masks have to "fight" to fill the corresponding pixels (to ensure the exactly one class per pixel). But from your raw mask plot it seems that you have a good front door mask, it's just unfortunate that the confidence is low
Here are a bunch of suggestions, some of which are mutually exclusive, while other are additive, feel free to pick what makes sense:

- Try to visualize the results you get with the segmentation post-processor (https://github.com/facebookresearch/detr/blob/4e1a9281bc5621dcd65f3438631de25e255c4269/models/segmentation.py#L218). You can play with the binary mask threshold, and possibly also manually threshold detections you want to keep based on confidence. This should bring results that are closer to mask-RCNN, but with sharper boundaries hopefully.
- To be able to reliably pick a detection threshold (no matter which post-processor you use), it might be necessary to calibrate the model's confidence on an held-out validation set (preferably real-life images and not synthetic). As an example, the model seems to be extremely certain about the "not-door" class (which I assume to be present in virtually all training images), but it's possible that it struggles a bit more with specific doors, esp if they don't always appear).
- A heuristic you can apply to both the segmentation and the panoptic post-processors is to prevent the model from predicting a mask outside of its corresponding box (you simply clip it to the boundaries of the box). This would prevent artifacts like the one you showed in the very first picture


Best of luck
